### PR TITLE
POSIX extensions: corelibc headers

### PR DESCRIPTION
**Pull request by mikbras**
This PR is the first of many for the POSIX extensions feature (formerly the I/O feature). 

This particular PR extends the corelibc headers with definitions for the forthcoming POSIX capabilities. It only defines interface and has no effect on existing functionality. Later PRs will implement this interface.

In keeping with our corelibc policy, we do not document any standard C or POSIX capabilities, as we do not wish to reproduce standard documentation here. For more information, please refer to the POSIX documentation.

My responses are a bit scattered, so here's a high-level summary:

- Note that corelibc is an internal library. It is used by to build our internal mbedtls, openssl, and libc.
- End-users will use MUSL libc (not corelibc), which redirects inernally to corelibc in various cases.
- This PR exposes functions for sockets and files and other things but calls to create sockets and filese result in failure if there are no devices installed to handle these. For example, if you call **open()**, you get an error if there are no file system devices installed. By default, not devices are installed.
- All definitions provided by **corelibc** are also provided by MUSL libc and glibc (including sockets and files).

### CODE REVIEW

File: enclave/core/sgx/asmdefs.h [hunk 1]
```diff
@@ -38,9 +38,9 @@
 #define td_callsites (td_oret_arg + 8)
 #define td_simulate (td_callsites + 8)
 
-#define oe_exit __morestack
+#define oe_exit_enclave __morestack
 #ifndef __ASSEMBLER__
-void oe_exit(uint64_t arg1, uint64_t arg2);
```
Comment (mikbras) [inline enclave/core/sgx/asmdefs.h]: Note to reviewers: oe_exit() was renamed to avoid a conflict with the one in stdlib.h.

File: include/openenclave/corelibc/assert.h [hunk 2]
```diff
@@ -4,11 +4,30 @@
...
+** Standard-C names:
+**
+**==============================================================================
+*/
+
+#if defined(OE_NEED_STDC_NAMES)
+
+#define assert oe_assert
+
+#define __assert_fail __oe_assert_fail
```
Comment (anitagov) [inline include/openenclave/corelibc/assert.h]: Curious why this change is needed.

File: include/openenclave/corelibc/arpa/inet.h [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+{
+    return oe_ntohs(x);
+}
+
+OE_INLINE uint32_t htonl(uint32_t x)
+{
+    return oe_htonl(x);
+}
+
+OE_INLINE uint16_t htons(uint16_t x)
```
Comment (anakrish) [inline include/openenclave/corelibc/arpa/inet.h]: Just for my understanding. Why do these need to be inline functions and not 
#define htos oe_htons
?

File: include/openenclave/corelibc/bits/jmp_buf.h [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    uint64_t rip;
+    uint64_t rbx;
+    uint64_t r12;
+    uint64_t r13;
+    uint64_t r14;
+    uint64_t r15;
+    /* Added these to align with size of MUSL jmp_buf */
+    uint64_t __fl;
+    uint64_t __ss[128 / sizeof(long)];
+} __OE_JMP_BUF;
```
Comment (anitagov) [inline include/openenclave/corelibc/bits/jmp_buf.h]: Additionally, the control bits of the MXCSR register and the x87 control word are preserved across calls i.e. callee-saved. Let's add this info to the comments section.
Also the direction flag DF in the RFlags register must be clear (set to "forward" direction) on function entry and return i.e. preserved.

File: include/openenclave/corelibc/arpa/inet.h (hunk 3 above)
Comment (dthaler) [reply include/openenclave/corelibc/arpa/inet.h]: # define htons             oe_htons
is how it appears in the new_platforms branch, so I have the same question

File: enclave/core/sgx/asmdefs.h (hunk 1 above)
Comment (dthaler) [reply enclave/core/sgx/asmdefs.h]: I'd suggest adding a /* comment */ in the code explaining the difference

**Comment by paulcallen** [comment]
file and socket functionality is opt in at an API layer within the enclave. You enable the feature and then it can be used. By default the functionality is not enabled.

**Comment by anitagov** [comment]
Thanks @paulcallen . OK, the new .h files and libraries will get included only when the user opts in. Nice.

**Comment by paulcallen** [comment]
the headers themselves will be included, but the APIs will fail if the feature is not enabled

**Comment by anakrish** [comment]
> file and socket functionality is opt in at an API layer within the enclave. You enable the feature and then it can be used. By default the functionality is not enabled.

Should these go to **corelibc**? sockets/posix apis are not part of a **libc** library.
Maybe we should rename corelibc to something else more inclusive of  posix apis.

**Comment by sevandbhel** [comment]
does oecore.lib always contain socket and file functionality? So there aren’t specific libs for socket and files to link against

File: include/openenclave/corelibc/bits/addrinfo.h [hunk 5]
```diff
@@ -0,0 +1,14 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+struct __OE_ADDRINFO
+{
```
Comment (soccerGB) [inline include/openenclave/corelibc/bits/addrinfo.h]: For basic posix types like addrinfo  and oe_socklen_t , given we are enabling POSIX API, I am wondering what value it has by defining exact one to one OE_* matching type. I noticed direct castings are used when calling actual host side socket APIs.

File: include/openenclave/corelibc/netinet/in.h [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+#define INADDR_NONE OE_INADDR_NONE
+#define INADDR_LOOPBACK OE_INADDR_LOOPBACK
+#define SOCK_STREAM OE_SOCK_STREAM
+#define SOCK_DGRAM OE_SOCK_DGRAM
+#define IN6ADDR_ANY_INIT OE_IN6ADDR_ANY_INIT
+#define IN6ADDR_LOOPBACK_INIT OE_IN6ADDR_LOOPBACK_INIT
+#define s6_addr oe_s6_addr
+#define s6_addr16 oe_s6_addr16
+#define s6_addr32 oe_s6_addr32
+
```
Comment (soccerGB) [inline include/openenclave/corelibc/netinet/in.h]: This comment is similar to my previous one, I am wondering what the actual benefit of doing  above new defines.

File: enclave/core/sgx/asmdefs.h (hunk 1 above)
Comment (mikbras) [reply enclave/core/sgx/asmdefs.h]: Done. Thanks Dave.

File: include/openenclave/corelibc/assert.h (hunk 2 above)
Comment (mikbras) [reply include/openenclave/corelibc/assert.h]: These definitions were missing. Corelibc must provide both oe-prefixed names and standard-C names for everything. This change will allow programs such as mbedtls and OpenSSL to compile when calling assert().

File: include/openenclave/corelibc/arpa/inet.h (hunk 3 above)
Comment (mikbras) [reply include/openenclave/corelibc/arpa/inet.h]: In this case macros would work fine. In other cases where function parameters need to be cast to oe-prefixed types, inlines work better. This has been our convention before this PR.

Consider this case.

```
OE_INLINE int stat(const char* pathname, struct stat* buf)
{
    return oe_stat(pathname, (struct oe_stat*)buf);
}
```

One could replace this with:

```
#define stat oe_stat
```

But this is subtle, because "stat" is both a function name and a structure name. Worse, this substitution propagates into user code, like this.

```
struct stat stat; /* stat renamed to oe_stat. If this variable is not used, the compiler warning will read "oe_stat is unused". */
```

Also, you have to worry about user code that redefines "stat" after including the corelibc header.

Finally, the macro approach does not play well with C++ function overloading. Consider the following.

```
int oe_foo(int x);
#define foo oe_foo
.
.
.
/* User defined functions. */
int oe_foo(int x, int y)
{
    ...
}
int foo(int x, int y)
{
    ...
}
```

This results in a "oe_foo" already defined error.

File: include/openenclave/corelibc/bits/jmp_buf.h (hunk 4 above)
Comment (mikbras) [reply include/openenclave/corelibc/bits/jmp_buf.h]: We didn't write this code. We just moved it to a different header. This should be tracked (as a separate) as a separate issue.

**Comment by mikbras** [comment]
> file and socket functionality is opt in at an API layer within the enclave. You enable the feature and then it can be used. By default the functionality is not enabled.

**Comment by mikbras** [comment]
> does oecore.lib always contain socket and file functionality? So there aren’t specific libs for socket and files to link against

It contains the calls (socket(), open(), etc.) but it does not contain any devices to carry out the calls, so these will result in errnos. It is the same in Linux: the calls are defined but there may be no devices installed to handle these calls. For example, if the socket layer is not configured, socket() will fail.

**Comment by mikbras** [comment]
> file and socket functionality is opt in at an API layer within the enclave. You enable the feature and then it can be used. By default the functionality is not enabled.
> 
> Should these go to corelibc? sockets/posix apis are not part of a libc library.
> Maybe we should rename corelibc to something else more inclusive of  posix apis.

Anand: All of these new APIs are in MUSL libc as well as glibc. Linux considers these part of the C runtime.

**Comment by mikbras** [comment]
Accidentally closed.

**Comment by mikbras** [comment]
> Very high level question:
> Why are these headers part of "corelibc"?
> My understanding was that sockets, files, etc are optional since they increase the surface area for attacks.

mikbras> This exposes the interfaces with their standard names so that mbedtls and openssl can compile using these definitions. The features themselves are disabled until you load the corresponding module. For example:

```
/* Load the host sockets module. */
oe_load_module_hostsock();
```
Calling **socket()** fails if the module is not loaded with **ENOTSUP**.

**Comment by mikbras** [comment]
> does oecore.lib always contain socket and file functionality? So there aren’t specific libs for socket and files to link against

Yes, but the functionality is disabled if there are no devices installed to handle these calls. It is the same in Linux: if there are no devices installed, calls like open() and socket() fail.

File: include/openenclave/corelibc/bits/addrinfo.h (hunk 5 above)
Comment (mikbras) [reply include/openenclave/corelibc/bits/addrinfo.h]: Good question. We provide the oe-prefixed types so they can be passed to the host side in OCALLs. Without the oe-prefixed names, these would conflict with Windows definitions. For example, many of these integer typedefs have different sizes on Windows.

File: include/openenclave/corelibc/netinet/in.h (hunk 6 above)
Comment (mikbras) [reply include/openenclave/corelibc/netinet/in.h]: These are struct field defines. Linux uses these alot and we don't want to collide with their defintions. Our corelibc headers need to coexist with standard-C headers when OE_NEED_STDC_NAMES is not defined. For example, we want to use <openenclave/corelibc/netinet/in.h> on the host side so we can use oe-prefixed definitions sent over in OCALLs. Our structures have fields named s6_addr, so we renamed them to oe_s6_addr. Without this workaround, Windows provides it's own definition of s6_addr which refers to a different field name in their structure.

**Comment by mikbras** [comment]
> I have similar high-level question as Anand.
> Shouldn't the File I/O headers be user opt-in instead of the default?
> I also prefer the term I/O instead of POSIX for the feature as I/O is disallowed inside the SGX enclave.

The term POSIX only appears in the PR, not the code. Also I/O is not broad enough because the PR introduces some non-I/O related calls needed by mbedtls. These include getpid() and gethostname() among others.

**Comment by mikbras** [comment]
> file and socket functionality is opt in at an API layer within the enclave. You enable the feature and then it can be used. By default the functionality is not enabled.
> 
> Should these go to corelibc? sockets/posix apis are not part of a libc library.
> Maybe we should rename corelibc to something else more inclusive of  posix apis.

**Comment by mikbras** [comment]
bors try

**Comment by mikbras** [comment]
bors try

File: include/openenclave/corelibc/assert.h (hunk 2 above)
Comment (anitagov) [reply include/openenclave/corelibc/assert.h]: Great. Thanks for the insight.

File: include/openenclave/corelibc/bits/jmp_buf.h (hunk 4 above)
Comment (anitagov) [reply include/openenclave/corelibc/bits/jmp_buf.h]: ok, sounds good.

File: include/openenclave/corelibc/bits/addrinfo.h (hunk 5 above)
Comment (soccerGB) [reply include/openenclave/corelibc/bits/addrinfo.h]: Thanks for addressing my comment.

File: include/openenclave/corelibc/netinet/in.h (hunk 6 above)
Comment (soccerGB) [reply include/openenclave/corelibc/netinet/in.h]: Thanks for addressing my comment.

**Comment by anakrish** [comment]
> > file and socket functionality is opt in at an API layer within the enclave. You enable the feature and then it can be used. By default the functionality is not enabled.
> > Should these go to corelibc? sockets/posix apis are not part of a libc library.
> > Maybe we should rename corelibc to something else more inclusive of  posix apis.
> 
> These APIs are also in the MUSL API and glibc. Linux does not separate the C-standard from the POSIX standard. The C-standard defines are all part of the POSIX standard.

Ok. I have more of a C standard/compiler writer view of libc. Along the lines of http://www.cplusplus.com/reference/clibrary/. Hence my confusion about what goes into libc.

Given that libc is adding posix functionality, I would prefer that we rename "corelibc" to corepal (platform abstraction layer) or something more descriptive. 
Otherwise, it becomes hard to talk about oelibc, bring-your-own-libc, what belongs in oecore etc.

**Comment by mikbras** [comment]
bors try

**Comment by johnkord** [comment]
bors try

**Comment by johnkord** [comment]
bors try
